_posts/2021-08-14-Prefer-Scoped-Enums-Over-Unscoped-Enums.md

```diff
@@ -0,0 +1,210 @@
+---
+layout: post
+title: "Prefer Scoped Enums Over Unscoped Enums (Notes)"
+date: 2021-08-14
+desc: "Notes on Item 10, Chapter 3 of Effective Modern C++"
+keywords: "Blog, C++, Enumerators, Enums, Notes"
+categories: [Blog, C++]
+tags: [Blog, C++]
+blog: [C++]
+published: true
+excerpt_separator: <!--more-->
+show_sidebar: false
+author: "Kushashwa Ravi Shrimali"
+icon: icon-html
+---
+
+## Scoped vs Unscoped Enums 
+
+* **General rule:** declaring a name inside curly braces is limited to that scope.
+* **Exception:** C++-98 style Enums
+
+<!--more-->
+
+---
+**NOTE**
+
+My notes on Chapter 3, Item 10 of Effective Modern C++ written by Scott Meyers.
+
+Some (or even all) of the text can be similar to what you see in the book, as these are notes: I've tried not to be unnecessarily creative with my words. :)
+
+---
+
+```cpp
+// You can't declare black, white, red in the scope containing the enum Color
+enum Color {
+    black, white, red;
+};
+
+auto white = false; // error: white already declared in this scope
+```
+
+* Unscoped Enums have implicit type conversions for their enumerators.
+* Enumerators can implicitly convert to integral types, and then to floating-point types.
+
+```cpp
+// Assume Color is declared like above
+Color c = red; // valid since Enumerator white is leaded to the scope Color is in
+if (c < 10) {  // valid, implicit conversion
+    // ... do something
+}
+else if (c < 10.5) {  // also valid, implicit conversion
+    // ... do something
+}
+```
+
+The C++-9 Style Enums are termed as uncoped Enums (because of leaking names).
+
+**C++-11 Scoped Enums**:
+
+```cpp
+// black, white, red are now scoped to Color Enum
+enum class Color {
+    black, white, red;
+};
+
+// This is now valid
+auto white = false;
+```
+
+Separately, if you do: (consider `Color` Enum has already been declared)
+
+```cpp
+Color c = white; // error: no enumerator named "white" is in this scope
+Color c = Color::white; // valid
+auto c = Color::white; // valid
+```
+
+* Also referred as enum classes (because declared using `enum class`).
+* Enumerators in scoped Enums are strongly typed (no implicit type conversion)
+
+```cpp
+// Assume Color is declared as above using enum class
+Color c = Color::red;
+
+if (c < 10.5) {  // Error! can't compare Color and double
+    // do something...
+}
+```
+
+**Note:** you can do explicit casting using `cast`.
+**Note about enums in C++:**
+    * Every enum in C++ has an integral underlying type that is determined by compilers.
+    * Compilers need to know the size of enum before using it.
+
+## C++98 vs C++11 on Enums
+
+**C++98:**
+
+* Unscoped enums can not be forward-declared.
```

Comment:
Can be with hacks? As you have mentioned below? 
I might be wrong though.

Reply:
Not in C++-98, the compiler needs to know the values in the enumerator to know the storage of the enum.

I'll add these to good reads:

* Proposal for forward declaration to enums (accepted), dated 2008: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2764.pdf
* Forward Declaring an Enum in C++? https://stackoverflow.com/questions/71416/forward-declaring-an-enum-in-c

Reply:
Ah. I see. Thanks for the resources!